File: vscode-extension/runtime/tools/generate-report.mjs
#!/usr/bin/env node
// TRA report generator: zero external dependencies. Loads a TRA project folder,
// computes initial and residual risk, runs the plausibility checker, and writes report/index.html.
// Usage: node tools/generate-report.mjs <project-dir>
import { readFileSync, writeFileSync, mkdirSync, existsSync, readdirSync } from "node:fs";
import { isAbsolute, join, dirname } from "node:path";
import { fileURLToPath } from "node:url";

const root = dirname(dirname(fileURLToPath(import.meta.url)));
const hasTraProject = (dir) => existsSync(join(dir, "01-project-description/project.json"));

function resolveProjectPath() {
  const fromArg = process.argv[2] || process.env.EMBEDRISK_PROJECT_DIR;
  if (fromArg) {
    const full = isAbsolute(fromArg) ? fromArg : join(root, fromArg);
    if (!hasTraProject(full)) throw new Error(`TRA project not found at: ${full}`);
    return { display: fromArg, base: full };
  }

  for (const searchRoot of [join(root, "projects"), join(root, "../../webapp/projects")]) {
    if (!existsSync(searchRoot)) continue;
    for (const entry of readdirSync(searchRoot, { withFileTypes: true })) {
      if (!entry.isDirectory()) continue;
      const full = join(searchRoot, entry.name);
      if (hasTraProject(full)) return { display: full, base: full };
    }
  }

  throw new Error("No TRA project found. Pass a project path as argv[2] or EMBEDRISK_PROJECT_DIR.");
}

const { display: proj, base } = resolveProjectPath();
const read = (p) => JSON.parse(readFileSync(join(base, p), "utf8"));
const esc = (s) => String(s ?? '').replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;').replace(/"/g, '&quot;');
const ID_PATTERN = /^[A-Za-z0-9 _\-:.]+$/;
const assumptionAnchor = (id) => `assumption-${String(id ?? '').replace(/[^A-Za-z0-9_.:-]/g, '-')}`;
const scheme = JSON.parse(readFileSync(join(root, ".assets/knowledge-base/risk-scheme.json"), "utf8"));

const project = read("01-project-description/project.json");
const ass = read(project.steps["02-assumptions"]);
const sys = read(project.steps["03-system-assets"]);
const threats = read(project.steps["06-threats"]).threats;
const cms = read(project.steps["08-countermeasures"]).countermeasures;
let useCases = { diagrams: [] };
try { useCases = read(project.steps["04b-use-cases"] || "04b-use-cases/use-cases.json"); } catch { }
let reqs = [];
try { reqs = read(project.steps["05-requirements"]).requirements || []; } catch { }

const band = (r) => scheme.matrix.bands.find((b) => r >= b.min && r <= b.max) || scheme.matrix.bands[0];
const residual = (t) => {
  const links = cms.flatMap((c) => c.addresses.filter((a) => a.threat === t.id));
  if (!links.length) return { l: t.likelihood, i: t.impact };
  return { l: Math.min(...links.map((a) => a.residualLikelihood ?? t.likelihood)), i: Math.min(...links.map((a) => a.residualImpact ?? t.impact)) };
};

// Plausibility checker
const ids = new Set(sys.components.map((c) => c.id));
const tids = new Set(threats.map((t) => t.id));
const assumptionRows = [
  ...(ass.device || []).map((a) => ({ id: a.id, text: a.text || '', source: 'Device' })),
  ...(ass.system || []).map((a) => ({ id: a.id, text: a.text || '', source: 'System' })),
  ...(ass.environment || []).map((a) => ({ id: a.id, text: a.text || '', source: 'Environment' })),
  ...(ass.operational || []).map((a) => ({ id: a.id, text: a.text || '', source: 'Operational' })),
  ...(ass.attacker || []).map((a) => ({ id: a.id, text: a.text || a.name || '', source: 'Attacker' })),
];
const assumptionIds = new Set(assumptionRows.map((a) => a.id).filter(Boolean));
const assumptionById = new Map(assumptionRows.map((a) => [a.id, a]));
const assumptionCitedBy = new Map();
const issues = [];
if (!ass.attacker?.length) issues.push("No attacker assumptions (required for likelihood).");
threats.forEach((t) => {
  if (!t.likelihood || !t.impact) issues.push(`${t.id}: missing risk rating.`);
  if (!t.components?.length) issues.push(`${t.id}: orphaned threat (no component).`);
  t.components?.forEach((c) => { if (!ids.has(c)) issues.push(`${t.id}: references unknown component ${c}.`); });
  (t.assumptionRefs || []).forEach((aid) => {
    if (!ID_PATTERN.test(aid)) issues.push(`${t.id}: references invalid assumption ID ${aid}.`);
    else if (!assumptionIds.has(aid)) issues.push(`${t.id}: references unknown assumption ${aid}.`);
    const refs = assumptionCitedBy.get(aid) || [];
    refs.push(t.id);
    assumptionCitedBy.set(aid, refs);
  });
});
cms.forEach((c) => c.addresses.forEach((a) => { if (!tids.has(a.threat)) issues.push(`${c.id}: addresses unknown threat ${a.threat}.`); }));

const rows = threats.map((t) => {
  const r = t.likelihood * t.impact, rr = residual(t), rrv = rr.l * rr.i;
  const refs = (t.assumptionRefs || []).map((aid) => {
    if (!ID_PATTERN.test(aid)) return `<span>${esc(aid)}</span> (invalid ID)`;
    const a = assumptionById.get(aid);
    if (!a) return `<span>${esc(aid)}</span> (missing)`;
    return `<a href="#${esc(assumptionAnchor(aid))}">${esc(aid)}</a> (${esc(a.source)})`;
  }).join(', ');
  const rationale = [t.likelihoodRationale ? `L: ${esc(t.likelihoodRationale)}` : '', t.impactRationale ? `I: ${esc(t.impactRationale)}` : ''].filter(Boolean).join('<br>');
  const classification = [
    `<b>Classification:</b> ${esc(t.classification || '—')}`,
    `<b>Responsibility:</b> ${esc(t.responsibility || '—')}`,
    t.deploymentConstraints ? `<b>Deployment constraints:</b> ${esc(t.deploymentConstraints)}` : '',
  ].filter(Boolean).join('<br>');
  return `<tr id="threat-${esc(t.id)}"><td>${esc(t.id)}</td><td>${esc(t.title)}</td><td>${esc((t.stride || []).join(''))}</td>` +
    `<td>${refs || '—'}</td><td>${rationale || '—'}</td><td>${classification}</td>` +
    `<td>${t.likelihood}×${t.impact}=<b style="color:${band(r).color}">${r} ${esc(band(r).name)}</b></td>` +
    `<td>${rr.l}×${rr.i}=<b style="color:${band(rrv).color}">${rrv} ${esc(band(rrv).name)}</b></td>` +
    `<td>${esc(t.status || 'open')}${t.status === 'accepted' && t.acceptedBy ? ` · <i>${esc(t.acceptedBy)}</i>` : ''}</td></tr>`;
}).join('');

const cmRows = cms.map((c) => {
  const tickets = (c.ticketUrls || []).filter(Boolean);
  return `<li><b>${esc(c.id)}</b> ${esc(c.title)} <i>(${esc(c.type || '')}, ${esc(c.status || '')})</i>` +
    ` → ${esc((c.addresses || []).map(a => a.threat).join(', '))}${tickets.map((u, i) => ` · <a href="${esc(u)}">ticket ${i + 1}</a>`).join('')}</li>`;
}).join('');

const reqRows = reqs.map((r) =>
  `<tr><td><b>${esc(r.id)}</b></td><td>${esc(r.text)}</td><td>${esc(r.standardRef || '')}${r.slFr ? ` · ${esc(r.slFr)}` : ''}</td>` +
  `<td>${esc((r.derivedFromThreat || []).join(', '))}</td><td>${esc((r.satisfiedByCM || []).join(', '))}</td></tr>`
).join('');
const useCaseRows = (useCases.diagrams || []).map((d) => `<li><b>${esc(d.name || d.id)}</b> — ${(d.entities || []).length} entities, ${(d.connections || []).length} connections</li>`).join('');
const assRows = ['device', 'system', 'environment', 'operational'].map((k) => {
  const items = ass[k] || [];
  if (!items.length) return '';
  return `<h3>${k[0].toUpperCase() + k.slice(1)} assumptions</h3><ul>${items.map((a) => {
    const cited = (assumptionCitedBy.get(a.id) || []).map((tid) => `<a href="#threat-${esc(tid)}">${esc(tid)}</a>`).join(', ');
    return `<li id="${esc(assumptionAnchor(a.id))}"><b>${esc(a.id)}</b> — ${esc(a.text || '')}${cited ? `<br><span style="font-size:11px;color:#555">cited by: ${cited}</span>` : ''}</li>`;
  }).join('')}</ul>`;
}).join('');
const atkRows = (ass.attacker || []).map((a) => {
  const cited = (assumptionCitedBy.get(a.id) || []).map((tid) => `<a href="#threat-${esc(tid)}">${esc(tid)}</a>`).join(', ');
  return `<li id="${esc(assumptionAnchor(a.id))}"><b>${esc(a.id)}</b> · <b>${esc(a.name)}</b> cap ${a.capability}, ${esc(a.access)}: ${esc(a.text || '')}${cited ? `<br><span style="font-size:11px;color:#555">cited by: ${cited}</span>` : ''}</li>`;
}).join('');

// SBOM (CycloneDX) from component list
const sbomMode = project.sbom?.mode || 'in-tool';
const sbomFormat = project.sbom?.format || 'cyclonedx';
const sbomComps = (sys.components || []).filter(c => c.kind !== 'external-entity');

const buildCdx = () => ({
  bomFormat: 'CycloneDX', specVersion: '1.5', version: 1,
  metadata: { timestamp: new Date().toISOString(), component: { type: 'device', name: project.device?.name || project.title || 'device', version: project.device?.version || '' } },
  components: sbomComps.map(c => ({
    'bom-ref': c.id, type: c.kind === 'hardware' ? 'device' : 'application',
    name: c.name, version: c.version || '',
    ...(c.supplier ? { supplier: { name: c.supplier } } : {}),
    ...(c.license ? { licenses: [{ license: { id: c.license } }] } : {}),
    ...(c.cpe && c.cpe.startsWith('cpe:') ? { cpe: c.cpe } : {}),
    ...(c.cpe && c.cpe.startsWith('pkg:') ? { purl: c.cpe } : {}),
    properties: [{ name: 'tra:provenance', value: c.provenance || 'unspecified' }],
  })),
});

const buildSpdx = () => {
  const sid = id => 'SPDXRef-Package-' + String(id).replace(/[^A-Za-z0-9.\-]/g, '-');
  const created = new Date().toISOString();
  const name = (project.device?.name || project.title || 'device').replace(/[^A-Za-z0-9.\-]/g, '-');
  return {
    spdxVersion: 'SPDX-2.3', dataLicense: 'CC0-1.0', SPDXID: 'SPDXRef-DOCUMENT',
    name: `${name}-sbom`, documentNamespace: `https://spdx.org/spdxdocs/${name}-${created}`,
    creationInfo: { created, creators: ['Tool: embedrisk'] },
    packages: sbomComps.map(c => {
      const externalRefs = [];
      if (c.cpe && c.cpe.startsWith('cpe:')) externalRefs.push({ referenceCategory: 'SECURITY', referenceType: 'cpe23Type', referenceLocator: c.cpe });
      if (c.cpe && c.cpe.startsWith('pkg:')) externalRefs.push({ referenceCategory: 'PACKAGE-MANAGER', referenceType: 'purl', referenceLocator: c.cpe });
      return { SPDXID: sid(c.id), name: c.name, versionInfo: c.version || 'NOASSERTION', supplier: c.supplier ? `Organization: ${c.supplier}` : 'NOASSERTION', downloadLocation: 'NOASSERTION', licenseConcluded: c.license || 'NOASSERTION', licenseDeclared: c.license || 'NOASSERTION', copyrightText: 'NOASSERTION', ...(externalRefs.length ? { externalRefs } : {}) };
    }),
    relationships: sbomComps.map(c => ({ spdxElementId: 'SPDXRef-DOCUMENT', relationshipType: 'DESCRIBES', relatedSpdxElement: sid(c.id) })),
  };
};

const sbomTableRows = sbomComps.map(c =>
  `<tr><td>${esc(c.name)}</td><td>${esc(c.version || '')}</td><td>${esc(c.supplier || '')}</td><td>${esc(c.license || '')}</td><td><code>${esc(c.cpe || '')}</code></td><td>${esc(c.provenance || 'own')}</td></tr>`
).join('');

let sbomSection;
if (sbomMode !== 'in-tool') {
  sbomSection = project.sbom?.url
    ? `<h2>Software bill of materials (SBOM)</h2><p>Provided externally: <a href="${esc(project.sbom.url)}">${esc(project.sbom.url)}</a></p>`
    : `<h2>Software bill of materials (SBOM)</h2><p>Declared as externally provided; no URL recorded.</p>`;
} else {
  const fmt = sbomFormat === 'spdx' ? 'SPDX 2.3' : 'CycloneDX';
  const file = sbomFormat === 'spdx' ? 'report/sbom.spdx.json' : 'report/sbom.cdx.json';
  sbomSection = sbomTableRows
    ? `<h2>Software bill of materials (SBOM)</h2><p>${fmt} — <code>${file}</code></p><table><tr><th>Component</th><th>Version</th><th>Supplier</th><th>License</th><th>CPE / purl</th><th>Provenance</th></tr>${sbomTableRows}</table>`
    : `<h2>Software bill of materials (SBOM)</h2><p>No components defined (add in step 03).</p>`;
}

const html = `<!doctype html><meta charset=utf8><title>TRA ${esc(project.title || project.device?.name || 'Report')}</title>
<style>body{font:14px system-ui;margin:2rem;max-width:72rem}table{border-collapse:collapse;width:100%}td,th{border:1px solid #ccc;padding:6px;text-align:left}th{background:#f5f5f5;font-size:12px;text-transform:uppercase;letter-spacing:.04em}h1{font-size:1.4rem}h2{font-size:1.1rem;margin-top:2rem}.warn{color:#c62828}code{font-size:12px}ul{padding-left:1.4rem}</style>
<h1>${esc(project.title || project.device?.name || 'TRA Report')}</h1>
<p>Device: <b>${esc(project.device?.name || '—')}</b> | SL-C ${esc(project.slTarget || '—')} | scope ${esc(project.scope?.mode || '—')} | boundary: ${esc(project.scope?.boundary || '—')}</p>
<h2>Attacker profiles</h2><ul>${atkRows}</ul>
${assRows ? `<h2>Assumptions</h2>${assRows}` : ''}
<h2>Assets</h2><ul>${(sys.assets || []).map(a => `<li>${esc(a.name)} (C${a.objectives?.confidentiality}/I${a.objectives?.integrity}/A${a.objectives?.availability}/S${a.objectives?.safety})</li>`).join('')}</ul>
<h2>Threats &amp; risk</h2><table><tr><th>ID</th><th>Threat</th><th>STRIDE</th><th>Assumptions</th><th>Rationale</th><th>Classification</th><th>Initial risk</th><th>Residual risk</th><th>Status</th></tr>${rows}</table>
<h2>Countermeasures</h2><ul>${cmRows || '<li>None defined.</li>'}</ul>
${reqRows ? `<h2>Security requirements</h2><table><tr><th>ID</th><th>Requirement</th><th>Standard ref</th><th>From threats</th><th>Satisfied by</th></tr>${reqRows}</table>` : ''}
${project.reportOptions?.includeUseCases ? `<h2>Use-case diagrams</h2>${useCaseRows ? `<ul>${useCaseRows}</ul>` : '<p>No use-case diagrams defined.</p>'}` : ''}
${sbomSection}
<h2>Plausibility check</h2>${issues.length ? `<ul class=warn>${issues.map(i => `<li>${esc(i)}</li>`).join('')}</ul>` : '<p>No issues found.</p>'}`;

mkdirSync(join(base, "report"), { recursive: true });
writeFileSync(join(base, "report/index.html"), html);
if (sbomMode === 'in-tool') {
  if (sbomFormat === 'spdx') {
    writeFileSync(join(base, "report/sbom.spdx.json"), JSON.stringify(buildSpdx(), null, 2) + '\n');
  } else {
    writeFileSync(join(base, "report/sbom.cdx.json"), JSON.stringify(buildCdx(), null, 2) + '\n');
  }
}
console.log(`Report written to ${join(proj, "report/index.html")}; ${issues.length} plausibility issue(s).`);

File: vscode-extension/runtime/tools/generate_report.py
#!/usr/bin/env python3
"""TRA report generator (zero dependencies). Mirrors tools/generate-report.mjs.
Usage: python tools/generate_report.py <project-dir>"""
import json, os, sys
from pathlib import Path

root = Path(__file__).resolve().parents[1]

def has_tra_project(d: Path) -> bool:
    return (d / "01-project-description/project.json").exists()

def resolve_project_path():
    arg = sys.argv[1] if len(sys.argv) > 1 else os.environ.get("EMBEDRISK_PROJECT_DIR")
    if arg:
        p = Path(arg) if os.path.isabs(arg) else (root / arg)
        if not has_tra_project(p):
            raise FileNotFoundError(f"TRA project not found at: {p}")
        return arg, p

    for search_root in (root / "projects", root / ".." / ".." / "webapp" / "projects"):
        if not search_root.exists():
            continue
        for entry in search_root.iterdir():
            if entry.is_dir() and has_tra_project(entry):
                return str(entry), entry

    raise FileNotFoundError("No TRA project found. Pass a project path as argv[1] or EMBEDRISK_PROJECT_DIR.")

proj, base = resolve_project_path()
read = lambda p: json.loads((base / p).read_text(encoding="utf8"))
scheme = json.loads((root / ".assets/knowledge-base/risk-scheme.json").read_text(encoding="utf8"))

project = read("01-project-description/project.json")
ass = read(project["steps"]["02-assumptions"])
sys_def = read(project["steps"]["03-system-assets"])
threats = read(project["steps"]["06-threats"])["threats"]
cms = read(project["steps"]["08-countermeasures"])["countermeasures"]
try:
    dfd = read(project["steps"]["04-dfd"])
except Exception:
    dfd = {"nodes": [], "flows": []}

def dfd_svg(dfd, layer=1):
    ROW = {"external-entity": 0, "process": 1, "multiprocess": 1, "store": 2}
    nodes = [n for n in dfd["nodes"] if n.get("layer") == layer and n["type"] != "trust-boundary"]
    col = {0: 0, 1: 0, 2: 0}
    pos = {}
    out = []
    for n in nodes:
        r = ROW.get(n["type"], 1); x = 30 + col[r] * 180; y = 50 + r * 110; col[r] += 1
        pos[n["id"]] = (x, y)
        shape = (f"<ellipse cx='{x+60}' cy='{y+25}' rx='60' ry='25' fill='#fff' stroke='#333'/>" if n["type"] in ("process", "multiprocess")
                 else f"<rect x='{x}' y='{y}' width='120' height='50' fill='none' stroke='#333' stroke-width='1' style='border-top:0'/>" if n["type"] == "store"
                 else f"<rect x='{x}' y='{y}' width='120' height='50' fill='#fff' stroke='#333'/>")
        out.append(shape + f"<text x='{x+60}' y='{y+28}' text-anchor='middle' font-size='11'>{n['label'][:18]}</text>")
    for f in dfd["flows"]:
        if f["from"] in pos and f["to"] in pos:
            x1, y1 = pos[f["from"]]; x2, y2 = pos[f["to"]]
            out.append(f"<line x1='{x1+60}' y1='{y1+25}' x2='{x2+60}' y2='{y2+25}' stroke='#666' marker-end='url(#a)'/><text x='{(x1+x2)/2+60}' y='{(y1+y2)/2+20}' font-size='9' fill='#666'>{f.get('label','')}</text>")
    return (f"<svg width='720' height='400' xmlns='http://www.w3.org/2000/svg'><defs><marker id='a' markerWidth='8' markerHeight='8' refX='7' refY='3' orient='auto'><path d='M0,0L7,3L0,6' fill='#666'/></marker></defs>{''.join(out)}</svg>")

def band(r):
    return next((b for b in scheme["matrix"]["bands"] if b["min"] <= r <= b["max"]), scheme["matrix"]["bands"][0])

def residual(t):
    links = [a for c in cms for a in c["addresses"] if a["threat"] == t["id"]]
    if not links:
        return t["likelihood"], t["impact"]
    # Single most-protective control (lowest residual product). Never combine the best likelihood
    # from one control with the best impact from another (Mauw & Oostdijk 2005; ISO/IEC 27005).
    best, best_score = (t["likelihood"], t["impact"]), None
    for a in links:
        rl = a.get("residualLikelihood", t["likelihood"]); ri = a.get("residualImpact", t["impact"])
        s = (rl or 0) * (ri or 0)
        if best_score is None or s < best_score:
            best_score, best = s, (rl, ri)
    return best

ids = {c["id"] for c in sys_def["components"]}
tids = {t["id"] for t in threats}
tById = {t["id"]: t for t in threats}
issues = []
if not ass.get("attacker"):
    issues.append("No attacker assumptions (required for likelihood).")
for t in threats:
    if not t.get("likelihood") or not t.get("impact"):
        issues.append(f"{t['id']}: missing risk rating.")
    if not t.get("stride"):
        issues.append(f"{t['id']}: no STRIDE category assigned.")
    if not t.get("components"):
        issues.append(f"{t['id']}: orphaned threat (no component).")
    for c in t.get("components", []):
        if c not in ids:
            issues.append(f"{t['id']}: references unknown component {c}.")
    if t.get("status") == "mitigated":
        impl = any(a["threat"] == t["id"] and c.get("status") in ("implemented", "verified") for c in cms for a in c["addresses"])
        if not impl:
            issues.append(f"{t['id']}: status 'mitigated' but no implemented/verified countermeasure addresses it.")
for c in cms:
    for a in c["addresses"]:
        if a["threat"] not in tids:
            issues.append(f"{c['id']}: addresses unknown threat {a['threat']}.")
            continue
        th = tById[a["threat"]]
        if a.get("residualImpact") is not None and a["residualImpact"] > th["impact"]:
            issues.append(f"{c['id']}: residual impact for {a['threat']} ({a['residualImpact']}) exceeds initial ({th['impact']}).")
        if a.get("residualLikelihood") is not None and a["residualLikelihood"] > th["likelihood"]:
            issues.append(f"{c['id']}: residual likelihood for {a['threat']} ({a['residualLikelihood']}) exceeds initial ({th['likelihood']}).")
        if c.get("type") == "preventive" and a.get("residualImpact") is not None and a["residualImpact"] < th["impact"]:
            issues.append(f"{c['id']}: a preventive control should reduce likelihood, not impact ({a['threat']}).")

# Lightweight JSON-Schema validation (zero dependency): required keys, types, enums, ranges.
_T = {"string": str, "integer": int, "number": (int, float), "object": dict, "array": list, "boolean": bool}
def _check(inst, sch, path, out):
    t = sch.get("type")
    if isinstance(t, str) and t in _T and not isinstance(inst, _T[t]):
        out.append(f"{path or 'root'}: expected {t}"); return out
    if "enum" in sch and inst not in sch["enum"]:
        out.append(f"{path}: '{inst}' not allowed")
    if isinstance(inst, dict):
        for k in sch.get("required", []):
            if k not in inst: out.append(f"{path}/{k} required")
        for k, sub in sch.get("properties", {}).items():
            if k in inst: _check(inst[k], sub, f"{path}/{k}", out)
    if isinstance(inst, list) and "items" in sch:
        for i, it in enumerate(inst): _check(it, sch["items"], f"{path}[{i}]", out)
    if isinstance(inst, (int, float)) and not isinstance(inst, bool):
        if "minimum" in sch and inst < sch["minimum"]: out.append(f"{path}: below minimum")
        if "maximum" in sch and inst > sch["maximum"]: out.append(f"{path}: above maximum")
    return out
_sd = root / ".assets/schema"
for inst, sname in [(project, "tra-project"), (ass, "assumptions"), (sys_def, "system"), (dfd, "dfd"), ({"threats": threats}, "threats"), ({"countermeasures": cms}, "countermeasures")]:
    try:
        for e in _check(inst, json.loads((_sd / f"{sname}.schema.json").read_text(encoding="utf8")), "", [])[:8]:
            issues.append(f"schema [{sname}] {e}")
    except Exception as ex:
        issues.append(f"schema [{sname}] {ex}")

rows = ""
for t in threats:
    r = t["likelihood"] * t["impact"]; rl, ri = residual(t); rr = rl * ri
    rows += f"<tr><td>{t['id']}</td><td>{t['title']}</td><td>{''.join(t['stride'])}</td><td>{t['likelihood']}x{t['impact']}=<b style='color:{band(r)['color']}'>{r} {band(r)['name']}</b></td><td>{rl}x{ri}=<b style='color:{band(rr)['color']}'>{rr} {band(rr)['name']}</b></td><td>{t['status']}</td></tr>"

assets = "".join(f"<li>{a['name']} (C{a['objectives']['confidentiality']}/I{a['objectives']['integrity']}/A{a['objectives']['availability']}/S{a['objectives']['safety']})</li>" for a in sys_def["assets"])
atk = "".join(f"<li><b>{a['name']}</b> cap {a['capability']}, {a['access']}: {a['text']}</li>" for a in ass["attacker"])
cmrows = "".join(f"<li><b>{c['id']}</b> {c['title']} → {', '.join(a['threat'] for a in c['addresses'])} ({c.get('status','')})</li>" for c in cms)

def _esc(s):
    return str(s).replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")

def _tree_ul(node):
    kids = node.get("children") or []
    structural = node.get("kind") in ("goal", "step", "substep", "category")
    meta = f" <em>[{node.get('gate', 'AND')}]</em>" if (structural and kids) else ""
    if node.get("kind") == "countermeasure":
        meta += f" <span style='color:#2e7d32'>(defence{(' ' + node['countermeasureRef']) if node.get('countermeasureRef') else ''})</span>"
    if node.get("kind") == "vulnerability":
        meta += " <span style='color:#c62828'>(vulnerability)</span>"
    inner = "<ul>" + "".join(f"<li>{_tree_ul(c)}</li>" for c in kids) + "</ul>" if kids else ""
    return _esc(node.get("label", "")) + meta + inner

try:
    trees = read("07-attack-trees/attack-trees.json").get("trees", [])
except Exception:
    trees = []
trees_html = ""
for tr in trees:
    tref = tr.get("threatRef")
    if tref and tref not in tids:
        issues.append(f"Attack tree '{tr.get('id')}' references unknown threat '{tref}'.")
    heading = _esc(tr.get("title", tr.get("id", "")))
    if tref:
        heading += f" &mdash; {tref}"
    trees_html += f"<h3>{heading}</h3>{('<ul><li>' + _tree_ul(tr['root']) + '</li></ul>') if tr.get('root') else ''}"
trees_section = f"<h2>Attack trees</h2>{trees_html}" if trees_html else ""

warn = f"<ul class=warn>{''.join(f'<li>{i}</li>' for i in issues)}</ul>" if issues else "<p>No issues found.</p>"
html = (f"<!doctype html><meta charset=utf8><title>TRA {project['title']}</title>"
        "<style>body{font:14px system-ui;margin:2rem;max-width:60rem}table{border-collapse:collapse;width:100%}td,th{border:1px solid #ccc;padding:6px;text-align:left}h1{font-size:1.4rem}.warn{color:#c62828}</style>"
        f"<h1>{project['title']}</h1><p>Device: {project['device']['name']} | SL-T {project['slTarget']} | mode {project['scope']['mode']} | status {project['status']}</p>"
        f"<p><b>Scope:</b> {project['scope'].get('boundary','')}</p>"
        f"<h2>Data flow diagram (layer 1)</h2>{dfd_svg(dfd)}"
        f"<h2>Attacker profiles</h2><ul>{atk}</ul><h2>Assets</h2><ul>{assets}</ul>"
        f"<h2>Threats &amp; risk</h2><table><tr><th>ID</th><th>Threat</th><th>STRIDE</th><th>Initial</th><th>Residual</th><th>Status</th></tr>{rows}</table>"
        f"<h2>Countermeasures</h2><ul>{cmrows}</ul>"
        f"{trees_section}"
        f"<h2>Plausibility check</h2>{warn}")
(base / "report").mkdir(exist_ok=True)
(base / "report/index.html").write_text(html, encoding="utf8")
print(f"Report written to {proj}/report/index.html; {len(issues)} plausibility issue(s).")
